src/smt/smt_engine.cpp

```diff
@@ -3905,28 +3904,41 @@ void SmtEngine::declareSynthFun(const std::string& id,
                                 bool isInv,
                                 const std::vector<Expr>& vars)
 {
+  SmtScope smts(this);
+  finalOptionsAreSet();
+  doPendingPops();
```

Comment:
Do you need to do this now because of the attributes or because of the stale business maybe popping stuff? I don't remember well the dependencies.